### ISSUE TITLE
PriorityQueue replaces some elements by other elements

### ISSUE THREAD

**Issue by michitux** [stxxl/stxxl#20]
Unless I misunderstood the documentation of the priority queue I have found a bug either in the priority queue or in one of the underlying primitives it is using. I want to have a priority queue that has many elements which are all different but some of them are equal when comparing them, i.e. neither `a < b` nor `b < a` holds. As I did not find any contrary information I assume that this should be supported by the stxxl priority queue. Unfortunately I had to spend several hours debugging until I found out that it is apparently not supported, at least not if the queue is large (I'm not sure how large, but definitely larger than the internal memory it has). Here a small example which triggers the problem:

``` c++
#include <stxxl/random>
#include <stxxl/priority_queue>

struct myElement {
    stxxl::int64 sortKey;
    stxxl::int64 id;
};


inline std::ostream &operator<<(std::ostream &os, myElement const &m) {
    return os << "{sortKey:" << m.sortKey << ", id:" << m.id << "}";
}

int main() {
    stxxl::int64 numElements = 100 * 1000 * 1000;
    stxxl::int64 maxKey = 256;

    stxxl::random_number32 rand;

    struct MyCompareLess {
        bool operator()(const myElement &a, const myElement &b) const { return a.sortKey < b.sortKey; };
        myElement min_value() const { return {std::numeric_limits<stxxl::int64>::min(), std::numeric_limits<stxxl::int64>::min()}; };
    };


    typedef stxxl::PRIORITY_QUEUE_GENERATOR<myElement, MyCompareLess, 16*1024*1024, 1024*1024>::result pqueue_type;
    typedef pqueue_type::block_type block_type;
    stxxl::read_write_pool<block_type> pool(static_cast<stxxl::read_write_pool<block_type>::size_type>(8*1024*1024/block_type::raw_size),
                                            static_cast<stxxl::read_write_pool<block_type>::size_type>(8*1024*1024/block_type::raw_size));
    pqueue_type prioQueue(pool);

    // generate elements which have sometimes equal sorting keys but never equal ids.
    for (stxxl::int64 i = 0; i < numElements; ++i) {
        prioQueue.push({rand(maxKey), i});
    }

    std::vector<bool> idCheck(numElements, false);

    // check that no two elements that are removed from the queue have the same id.
    while (!prioQueue.empty()) {
        assert(!idCheck[prioQueue.top().id]); // this fails for some elements!
        idCheck[prioQueue.top().id] = true;
        prioQueue.pop();
    }

    return 0;
}
```

I'm using a debug build of STXXL version 1.4.1 and also checked it with the current master branch (same result) with the following disk configuration:

```
disk=/dev/null,12G,memory
```

Note that the assertion does not fail for many elements, just for a few hundred elements which appear to be random or maybe also evenly distributed.

**Comment by bingmann** [comment]
Yes, this is a very subtle problem with the PQ: the items have to be totally ordered.
The reason for this lies deeply in the way the PQ uses sentinels while merging groups. Therein it duplicates the sentinels, and atm it is unavoidable.

The fix for your case is simply:
```c++
bool operator()(const myElement &a, const myElement &b) const { 
    return (a.sortKey < b.sortKey) || (!(b.sortKey < a.sortKey) && (a.id < b.id));
};
```

**Comment by michitux** [comment]
If this is not considered as a bug in the STXXL itself, then I think it should at least be properly documented with a large warning that the comparison operator must define a **total ordering** and not as it is stated e.g. [here](http://stxxl.sourceforge.net/tags/master/design_pqueue.html) a **strict weak ordering** (quoting the page that is linked there: "A *total ordering* is one that satisfies an even stronger condition: equivalence must be the same as equality. ").

Is this requirement of a total ordering also present in other parts of the STXXL like regular sorters where according to the [documentation](http://stxxl.sourceforge.net/tags/master/design_algo_sort.html#StrictWeakOrdering) also a strict weak ordering should be sufficient?

**Comment by bingmann** [comment]
Hmm, yes it maybe a bug, but it cant be fixed without changing large parts of the merging code.
Maybe you want to provide a patch for the documentation?
